packages/stylist-macros/src/output/context_recorder.rs

```diff
@@ -1,37 +1,61 @@
 //! This module implements a type abstractly tracking in what kind of expression context
 //! an item appears. This information is leverage to provide improved performance and
 //! static caching of parts of the generated output.
-#[derive(Clone, Debug, PartialEq)]
-enum AllowedUsage {
+
+// The kind of an expression decribes in what context it can be used. It is harmless to
+// underapproximate the usage and e.g. classify expressions as Dynamic even if they are
+// actually Static.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum ExpressionKind {
+    // A dynamic expression refers to local variables and thus can not be used in global
+    // or compile-time `const` contexts.
     // ```
     // let width = 500;
     // style! { width: ${width}; }
-    // //               ^^^^^^ dynamic expression, can't wrap style in Lazy
+    // //                ^^^^^ dynamic expression, can't wrap style in Lazy
     // ```
     Dynamic,
+    // A static expression refers to global variables, which might still not be evaluatable
+    // in compile-time `const` contexts.
+    //
+    // E.g. assuming a library method `rgb` producing a string.
     // ```
-    // style! { width: 500px; }
-    // //       ------------- everything is static, do wrap style in Lazy
+    // style! { color: ${rgb(100, 100, 100)}; }
+    // //                ------------------ expression is static, can wrap style in Lazy
     // ```
     Static,
-    // TODO: we can probably avoid a few allocations if we track which parts
-    // of the ast can be constructed statically (with const methods), which is
-    // even stronger than constructing it in the global context in a Lazy.
-    // Should you decide to implement this, keep in mind to change Self::MAX
-    // and adjust the generation of cow-vec tokens. Also check the usages of
-    // MaybeStatic::statick if they can be upgraded to Const.
-    // Const,
+    // An expression that can be evaluated in `const` context.
+    // We can avoid a few allocations if we track which parts of the ast can be constructed
+    // statically (with const methods), which is even stronger than constructing it in
+    // the global context in a Lazy.
+    // ```
+    // style! { color: black; }
+    // //       ------------- everything can be constructed in a const context
+    // ```
+    Const,
+}
+
+impl ExpressionKind {
+    // Compute the jointly allowed usage of two expression that are used together.
+    pub fn joint_kind(l: ExpressionKind, r: ExpressionKind) -> ExpressionKind {
```

Comment:
Adjust function signature to`pub fn join(&self, other: Self) -> Self` like `eq` or `cmp`?